File: GreenBite-backend/food/utils/mealdb_mapper.py
# food/utils/mealdb_mapper.py
def _clean(s: str | None) -> str:
    return (s or "").strip()

def extract_ingredients(meal: dict) -> list[dict]:
    items = []
    for i in range(1, 21):
        ing = _clean(meal.get(f"strIngredient{i}"))
        meas = _clean(meal.get(f"strMeasure{i}"))

        # MealDB sometimes returns "" or None; skip empty ingredients
        if not ing:
            continue

        text = f"{meas} {ing}".strip() if meas else ing

        items.append({
            "ingredient": ing,
            "measure": meas,
            "text": text,   # e.g. "1 pound penne rigate"
        })
    return items

def parse_tags(meal: dict) -> list[str]:
    raw = _clean(meal.get("strTags"))
    if not raw:
        return []
    # "Pasta,Curry" -> ["Pasta", "Curry"]
    return [t.strip() for t in raw.split(",") if t.strip()]

def mealdb_to_mealdb_fields(meal: dict) -> dict:
    return {
        "mealdb_id": _clean(meal.get("idMeal")),
        "title": _clean(meal.get("strMeal")),
        "category": _clean(meal.get("strCategory")),
        "cuisine": _clean(meal.get("strArea")),
        "instructions": _clean(meal.get("strInstructions")),
        "thumbnail": _clean(meal.get("strMealThumb")),
        "tags": parse_tags(meal),
        "ingredients": extract_ingredients(meal),
    }

File: GreenBite-frontend/src/utils/mealdb.mapper.js
export function instructionsToSteps(instructions) {
  return (instructions || "")
    .split(/\r?\n/)          // split on new lines
    .map(s => s.trim())
    .filter(Boolean);
}

export function formatIngredient(i) {
  const measure = (i?.measure || "").trim();
  const name = (i?.name || "").trim();
  return [measure, name].filter(Boolean).join(" ");
}

export function buildSavePayloadFromMealDb(meal){
  const steps = (meal.instructions || "")
  .split(/\r?\n/)
  .map(s => s.trim())
  .filter(Boolean);

  return {
    recipe: meal.title,
    title: meal.title,
    cuisine: meal.cuisine,
    mealTime: "Dinner",
    ingredients: (meal.ingredients || []). map(i => 
    [i.measure, i.name].filter(Boolean).join(" ").trim()
    ), steps,
  };
}
export function mapMealDbFromBackend(meal) {
  if (!meal) return null;

  return {
    mealdb_id: meal.mealdb_id,
    title: meal.title,
    thumbnail: meal.thumbnail,
    category: meal.category,
    cuisine: meal.cuisine,
    instructions: meal.instructions,
    tags: meal.tags || [],
    ingredients: meal.ingredients || [], // includes {text}
  };
}